Доступ к бд с помощью Hibernate, скрипты для создания таблиц, вставки…

smolkinakate17 opened the pull request:
… в них данных, очистки данных. Написан генератор на питоне

smolkinakate17 committed:
Доступ к бд с помощью Hibernate, скрипты для создания таблиц, вставки в них данных, очистки данных. Написан генератор на питоне

KorobetskayaAA commented on src/main/java/org/example/statistic/manager/Dictionary.java:
```diff
@@ -1,36 +1,94 @@
+
+    public Dictionary(EntityManager entityManager) {
+        this.entityManager = entityManager;
+        this.paymentManager = new PaymentManager(entityManager);
```
Нарушение принципа инверсии зависимостей (dependency inversion). PaymentManager должен создаваться и передаваться извне, скорее всего как singleton, т.е. один экземпляр на всё приложение.

Но в данном случае PaymentManager вообще не нужен, см. комментарий ниже

KorobetskayaAA commented on src/main/java/org/example/statistic/manager/Dictionary.java:
```diff
@@ -1,36 +1,94 @@
+
+    private Optional<T> find(T value) {
+
+        if (value instanceof Category) {
```
Такая реализация Dictionary делает его бессмысленным. Проще было бы написать два отдельных справочника, чем в каждом методе написать класс. Идея в том, чтобы сервис работал для любого справочника. Например, захотите вы добавить валюты, чтобы ничего менять не пришлось, просто создать Dictionary<Currency>.

Не нужно проверять, какой тип у T. Передать класс в find можно из экземпляра с помощью .getClass()

KorobetskayaAA commented on src/main/java/org/example/statistic/manager/Dictionary.java:
```diff
@@ -1,36 +1,94 @@
+            transaction.rollback();
+            throw new ValueNotExistException(value);
+        }
+        if (value instanceof Category) {
```
Не нужно проверять существование связанных сущностей вручную. Надо правильно настроить связь, запретив каскадное удаление. Тогда при попытке удалить категорию или метод, у которых есть платежи, БД просто откажется это делать.

KorobetskayaAA commented on src/main/java/org/example/statistic/manager/PaymentManager.java:
```diff
@@ -1,15 +1,81 @@
+        return Optional.of(p);
+    }
+
+    public void add(Payment payment) {
```
Желательно вернуть созданную сущность.

KorobetskayaAA commented on src/main/java/org/example/statistic/manager/PaymentManager.java:
```diff
@@ -1,15 +1,81 @@
+
+    }
+
+    public void update(Payment payment) {
```
Желательно вернуть измененную сущность.

KorobetskayaAA commented on src/main/java/org/example/statistic/manager/PaymentManager.java:
```diff
@@ -1,15 +1,81 @@
+        transaction.commit();
+    }
+
+    public void delete(Payment payment) {
```
Желательно вернуть удаленную сущность, и сделать перегрузку для удаления просто по id.

KorobetskayaAA commented on src/main/java/org/example/statistic/user/UserStatistic.java:
```diff
@@ -1,71 +1,102 @@
+                                "where p2.paymentDateTime between :fromDate and :toDate)*100,2) as percent " +
+                                "from Payment p1 where p1.paymentDateTime between :fromDate and :toDate " +
+                                "group by p1.paymentMethod"
+                        , Tuple.class
```
Очень запутанная реализация через Tuple. Почему не объявить и не вернуть сразу класс с нужными полями? Зачем вручную переписывать поля?

KorobetskayaAA commented on src/main/java/org/example/statistic/user/UserStatistic.java:
```diff
@@ -1,71 +1,102 @@
+                                "join Category c on pc.pk.category.id=c.id " +
+                                "where p.paymentDateTime between :fromDate and :toDate " +
+                                "group by category",
+                        Tuple.class
```
Аналогично, лучше сразу вернуть нужный класс.

KorobetskayaAA reviewed (changes requested):
Необходимо исправить Dictionary и запросы статистики

smolkinakate17 committed:
Добавлена постраничный вывод с сортировкой Payment по amount и paymentDateTime. Добавлен метод Payment.addCategory и Payment.findCategory.
В тесты добавлены методы для проверки постраничного вывода. В методах по сохранению и удалению добавлена проверка сохранения и удаления связаных сущностей

smolkinakate17 committed:
Исправлены недочеты
Сделала два менеджера для категорий и методов
(по хорошему хотелось бы вообще одного для категорий, методов и платежей, но у платежей есть метод с поиском по дате и постраничный вывод. Также не поняла как в старом Dictionary брать айди для поиска, потому что там тип Т)
Добавила классы DTO для UserStatistic.getPaymentMethodSharesMap и UserStatistic.getCategorySharesMap
Изменила таблицу payment_category чтобы нельзя было удалить категорию и метод по которым есть платежи, в тестах проверила проброс исключения

KorobetskayaAA reviewed (approved):
Принято.
